// ui/forms/counter.jsx
// libs
/* @flow */

import React, { PropTypes } from 'react';
import classNames from 'classnames';
import styles from './counter.css';
import cssModules from 'react-css-modules';

// components
import Button from '../buttons';

const Counter = props => {
  const {decreaseAction, increaseAction, disabled, ...rest} = props;

  return (
    <div styleName="container">
      <div styleName="counter-prepend">
        <Button disabled={disabled} onClick={decreaseAction} className="fc-btn-counter" />
      </div>
      <input
        type="number"
        className={classNames('fc-counter__input', props.className, {disabled})}
        disabled={disabled}
        {...rest}
      />
      <div styleName="counter-append">
        <Button disabled={disabled} onClick={increaseAction} className="fc-btn-counter" />
      </div>
    </div>
  );
};

Counter.propTypes = {
  value: PropTypes.number,
  min: PropTypes.number,
  max: PropTypes.number,
  step: PropTypes.number,
  decreaseAction: PropTypes.func,
  increaseAction: PropTypes.func,
  className: PropTypes.string,
};

Counter.defaultProps = {
  value: 1,
  step: 1,
  min: 1,
  max: 100,
  disabled: false,
};

export default cssModules(Counter, styles);
