// modules/async-utils.js
import _ from 'lodash';
import { assoc } from 'sprout-data';
import { createAction } from 'redux-act';
import { api } from 'lib/api';

export function reducer(state = {}, action) {
  const kind = _.get(action, 'meta.kind');
  if (kind == 'async') {
    const { type, namespace } = action.meta;
    switch (type) {
      case 'started':
        return assoc(state,
          [namespace, 'inProgress'], true,
          [namespace, 'finished'], false,
          [namespace, 'failed'], false
        );
      case 'succeeded':
        return assoc(state,
          [namespace, 'inProgress'], false,
          [namespace, 'finished'], true,
          [namespace, 'failed'], false
        );
      case 'failed':
        return assoc(state,
          [namespace, 'inProgress'], false,
          [namespace, 'finished'], true,
          [namespace, 'failed'], true
        );
      default:
        return state;
    }
  }
  return state;
}

function createAsyncAction(namespace, type, payloadReducer) {
  const description = `${namespace.toUpperCase()}_${type.toUpperCase()}`;
  return createAction(description, payloadReducer, () => ({
    kind: 'async',
    namespace,
    type,
  }));
}

const isServer = typeof self == 'undefined';

export default function createAsyncActions(namespace, asyncCall, payloadReducer) {
  const actions = {
    started: createAsyncAction(namespace, 'started', payloadReducer),
    succeeded: createAsyncAction(namespace, 'succeeded', payloadReducer),
    failed: createAsyncAction(namespace, 'failed', payloadReducer),
  };

  /* eslint-disable consistent-return */

  // @TODO! think about cancelling request on client side
  // for example in navigate to product 1, then to 2, then back to 1 and there user will see
  // not last navigated product, but product for which response will be last

  const perform = (...args) => {
    return (dispatch, getState, jwt) => {
      const apiContext = {
        jwt,
        phoenix: api.addAuth(jwt),
        dispatch,
      };

      dispatch(actions.started(...args));
      return asyncCall.call(apiContext, ...args)
        .then(
          res => dispatch(actions.succeeded(res))
        ).catch(err => {
          console.error(err && err.stack);
          dispatch(actions.failed(err));
        });
    };
  };

  const lazyPerform = (...args) => {
    let promise;

    return (dispatch, getState) => {
      const asyncState = _.get(getState(), ['asyncActions', namespace]);
      if (asyncState && isServer && (asyncState.inProgress || asyncState.finished)) {
        return promise;
      }

      promise = dispatch(perform(...args));
      return promise;
    };
  };

  /* eslint-enable consistent-return */

  return {
    fetch: lazyPerform, // for backward compartibility
    perform: lazyPerform,
    ...actions,
  };
}
