// modules/auth.js

/* @flow weak */

import { createAction, createReducer } from 'redux-act';
import fetch from 'isomorphic-fetch';
import { phoenix } from 'lib/api';

// types

export type TUser = {name: String, email: String};

export type LoginPayload = {
  email: string,
  password: string,
  kind: string,
};

export type UserState = {
  err: ?String,
  current: ?TUser,
  isFetching: boolean,
};

export type SignUpPayload = {
  name: string,
  email: string,
  password: string,
};

export const setUser = createAction('USER_SET');
const authStart = createAction('USER_AUTH_START');
const authError = createAction('USER_AUTH_ERROR');

const headers = {'Content-Type': 'application/json;charset=UTF-8'};

export function signUp(payload: SignUpPayload) {
  return () => {
    console.log(payload);
    return phoenix.post('/v1/public/registrations/new', payload).then((data) => {
      console.log(data);
    });
  };
}

export function authenticate(payload: LoginPayload) {
  return dispatch => {
    dispatch(authStart());
    return fetch('/v1/public/login', {
      method: 'POST',
      body: JSON.stringify(payload),
      credentials: 'same-origin',
      headers,
    }).then(response => {
      if (response.status == 200 && response.headers.get('jwt')) {
        localStorage.setItem('jwt', response.headers.get('jwt'));
        return response.json();
      }
      throw new Error('Server error, try again later. Sorry for inconvenience :(');
    }).then((token: TUser) => {
      if (token.email && token.name) {
        localStorage.setItem('user', JSON.stringify(token));
        return dispatch(setUser(token));
      }
      throw new Error('Server error, try again later. Sorry for inconvenience :(');
    }).catch(reason => {
      dispatch(authError(reason));
      throw new Error(reason);
    });
  };
}

const initialState = {isFetching: false};

const reducer = createReducer({
  [setUser]: (state: UserState, user: TUser) => {
    return {
      ...state,
      current: user,
      isFetching: false,
    };
  },
  [authStart]: (state: UserState) => {
    return {
      ...state,
      err: null,
      isFetching: true,
    };
  },
  [authError]: (state: UserState, error: string) => {
    return {
      ...state,
      err: error,
      isFetching: false,
    };
  },
}, initialState);

export default reducer;
