// modules/cart.js

/* @flow weak */

import _ from 'lodash';
import { createAction, createReducer } from 'redux-act';

export const toggleCart = createAction('TOGGLE_CART');
export const addToCart = createAction('ADD_TO_CART', (productId, quantity) => [productId, quantity]);

type ProductInCart = {
  skuId: number;
  quantity: number;
};

type FormData = {
  isVisible: boolean;
  skus: Array<ProductInCart>;
};

function addToSkuList(items, id, quantity) {
  const found = _.find(items, {skuId: id});
  let result = [];
  if (_.isEmpty(found)) {
    const sku = {
      skuId: id,
      quantity,
    };
    result = items.concat([sku]);
  } else {
    const newFound = {...found, quantity: found.quantity + quantity};
    result = _.map(items, (item) => {
      return item.skuId === id ? newFound : item;
    });
  }
  return result;
}

const initialState: FormData = {
  isVisible: false,
  skus: [],
};

const reducer = createReducer({
  [toggleCart]: state => {
    const currentState = _.get(state, 'isVisible', false);
    return {
      ...state,
      isVisible: !currentState,
    };
  },
  [addToCart]: (state, [productId, quantity]) => {
    const skus = addToSkuList(state.skus, productId, quantity);
    return {
      ...state,
      skus,
    };
  },
}, initialState);

export default reducer;
