// modules/checkout.js

/* @flow weak */

import _ from 'lodash';
import {createAction, createReducer} from 'redux-act';
import { assoc } from 'sprout-data';
import createAsyncActions from './async-utils';
import { phoenix } from 'lib/api';
import { fetchCountry } from 'modules/countries';
import { updateCart } from 'modules/cart';

export const AddressKind = {
  SHIPPING: 0,
  BILLING: 1,
};

export type AddressKindType = number;

export const EditStages = {
  SHIPPING: 0,
  DELIVERY: 1,
  BILLING: 2,
};

export type EditStage = number;

export type ShippingAddress = {
  city?: string;
}

export type CheckoutState = {
  editStage: EditStage;
  shippingAddress: ShippingAddress;
  billingAddress: ShippingAddress;
};

export type BillingData = {
  cardNumber?: string;
  month?: string;
  year?: string;
}

export const setEditStage = createAction('CHECKOUT_SET_EDIT_STAGE');
export const setBillingData = createAction('CHECKOUT_SET_BILLING_DATA', (key, value) => [key, value]);

export const setAddressData = createAction('CHECKOUT_SET_ADDRESS_DATA', (kind, key, value) => [kind, key, value]);
export const extendAddressData = createAction('CHECKOUT_EXTEND_ADDRESS_DATA', (kind, props) => [kind, props]);

/* eslint-disable quotes, quote-props */

function _fetchShippingMethods() {
  return this.phoenix.get('/v1/my/cart/shipping-methods');
}

/* eslint-enable quotes, quote-props */

const shippingMethodsActions = createAsyncActions('shipping-methods', _fetchShippingMethods);

export const fetchShippingMethods = shippingMethodsActions.fetch;
export const toggleSeparateBillingAddress = createAction('CHECKOUT_TOGGLE_BILLING_ADDRESS');

export function initAddressData(kind: AddressKindType): Function {
  return (dispatch, getState) => {
    const state = getState();

    const addr = getState().cart.shippingAddress;

    const countries = state.countries.list;

    const usaCountry = _.find(countries, {alpha3: 'USA'});
    const countryDetails = state.countries.details[usaCountry && usaCountry.id] || {
      regions: [],
    };

    let uiAddressData = {
      country: usaCountry,
      state: countryDetails.regions[0],
    };

    if (addr && addr.region) {
      dispatch(fetchCountry(addr.region.countryId)).then(() => {
        const countryInfo = getState().countries.details[addr.region.countryId];

        uiAddressData = _.pick(addr, ['name', 'address1', 'address2', 'city', 'zip', 'phoneNumber']);
        uiAddressData.country = countryInfo;
        uiAddressData.state = _.find(countryInfo.regions, {id: addr.region.id});

        dispatch(extendAddressData(kind, uiAddressData));
      });
    } else {
      dispatch(extendAddressData(kind, uiAddressData));
    }
  };
}

export function saveShippingAddress(): Function {
  return (dispatch, getState) => {
    const addr = getState().checkout.shippingAddress;
    const payload = _.pick(addr, ['name', 'address1', 'address2', 'city', 'zip', 'phoneNumber']);
    payload.regionId = addr.state.id;
    payload.phoneNumber = String(payload.phoneNumber);
    payload.isDefault = false;

    return phoenix.post('/v1/my/cart/shipping-address', payload)
      .then(res => {
        dispatch(updateCart(res.result));
      });
  };
}

export function saveShippingMethod(): Function {
  return (dispatch, getState) => {
    const payload = {
      shippingMethodId: getState().cart.shippingMethod.id,
    };

    return phoenix.patch('/v1/my/cart/shipping-method', payload)
      .then(res => {
        dispatch(updateCart(res.result));
      });
  };
}

export function saveGiftCard(code: string): Function {
  return () => {
    const payload = { code, amount: 10000 };

    return phoenix.post('/v1/my/cart/payment-methods/gift-cards', payload)
      .then(res => {
        console.info('saved', res);
      });
  };
}

const initialState: CheckoutState = {
  editStage: EditStages.SHIPPING,
  shippingAddress: {},
  billingAddress: {},
  billingData: {},
  billingAddressIsSame: true,
  shippingMethods: [],
};

const reducer = createReducer({
  [setEditStage]: (state, editStage: EditStage) => {
    return {
      ...state,
      editStage,
    };
  },
  [setAddressData]: (state, [kind, key, value]) => {
    const ns = kind == AddressKind.SHIPPING ? 'shippingAddress' : 'billingAddress';
    return assoc(state,
      [ns, key], value
    );
  },
  [extendAddressData]: (state, [kind, props]) => {
    const ns = kind == AddressKind.SHIPPING ? 'shippingAddress' : 'billingAddress';
    return assoc(state,
      [ns], props
    );
  },
  [setBillingData]: (state, [key, value]) => {
    return assoc(state,
      ['billingData', key], value
    );
  },
  [shippingMethodsActions.fetchSucceeded]: (state, list) => {
    return {
      ...state,
      shippingMethods: list,
    };
  },
  [toggleSeparateBillingAddress]: state => {
    return assoc(state,
      ['billingAddressIsSame'], !state.billingAddressIsSame
    );
  },
}, initialState);

export default reducer;
