// modules/products.js
/* @flow */

import _ from 'lodash';
import { createAction, createReducer } from 'redux-act';

const productsFetchStated = createAction('PRODUCTS_FETCH_STARTED');
const productsFetchSucceded = createAction('PRODUCTS_FETCH_SUCCEDED');
const productsFetchFailed = createAction('PRODUCTS_FETCH_FAILED');

export type Product = {
  id: number;
  name: string;
  imageUrl: string;
  price: string;
  categories: Array<number>;
}

type FormData = {
  isFetching: boolean;
  fetched: boolean;
  list: Array<Product>;
}

const mockedData : Array<Product> = [
  {
    id: 1,
    name: 'Donkey',
    imageUrl: 'https://s3-us-west-2.amazonaws.com/fc-firebird-public/images/product/Granger_Three_Quarter.jpg',
    price: '55$',
    categories: [2],
  },
  {
    id: 2,
    name: 'Shark',
    imageUrl: 'https://s3-us-west-2.amazonaws.com/fc-firebird-public/images/product/Festival_Three_Quarter.jpg',
    price: '20$',
    categories: [2, 3],
  },
  {
    id: 3,
    name: 'Sharkling',
    imageUrl: 'https://s3-us-west-2.amazonaws.com/fc-firebird-public/images/product/Gold_Blue_Three_Quarter.jpg',
    price: '30$',
    categories: [3],
  },
  {
    id: 4,
    name: 'Duck',
    imageUrl: 'https://s3-us-west-2.amazonaws.com/fc-firebird-public/images/product/Emma_Top_Front.jpg',
    price: '40$',
    categories: [3, 4],
  },
  {
    id: 5,
    name: 'Duckling',
    imageUrl: 'https://s3-us-west-2.amazonaws.com/fc-firebird-public/images/product/Marley_Top_Front.jpg',
    price: '100$',
    categories: [4],
  },
  {
    id: 6,
    name: 'Chicken',
    imageUrl: 'https://s3-us-west-2.amazonaws.com/fc-firebird-public/images/product/Round_Readers_Top_Front.jpg',
    price: '60$',
    categories: [4, 2],
  },
  {
    id: 7,
    name: 'Fox',
    imageUrl: 'https://s3-us-west-2.amazonaws.com/fc-firebird-public/images/product/Quay_Side.jpg',
    price: '87$',
    categories: [4, 2, 3],
  },
];

export function fetchProducts(categoryId: ?number): Function {
  return dispatch => {
    dispatch(productsFetchStated());

    if (categoryId == undefined) {
      return dispatch(productsFetchSucceded(mockedData));
    }

    const result = _.filter(mockedData, (item) => _.includes(item.categories, categoryId));
    return dispatch(productsFetchSucceded(result));
  };
}

const initialState: FormData = {
  isFetching: false,
  fetched: false,
  list: [],
};

const reducer = createReducer({
  [productsFetchStated]: state => {
    return {
      ...state,
      isFetching: true,
      fetched: false,
    };
  },
  [productsFetchSucceded]: (state, payload) => {
    return {
      ...state,
      list: payload,
      isFetching: false,
      fetched: true,
    };
  },
  [productsFetchFailed]: (state, err) => {
    console.error(err);

    return {
      ...state,
      isFetching: false,
      fetched: true,
    };
  },
}, initialState);

export default reducer;
