// modules/product-details.js
/* @flow */

type Attribute = {
  t: string,
  v: any,
};

type Attributes = { [key:string]: Attribute };

type BakedProduct = {
  id: number,
  attributes: Attributes,
  activeFrom?: string,
  activeTo?: string,
};

type BakedSku = {
  code: string,
  attributes: Attributes,
  activeFrom?: string,
  activeTo?: string,
};

type Context = {
  name: string,
  attributes: { [key:string]: string },
};

export type ProductResponse = {
  id: number,
  context: Context,
  product: BakedProduct,
  skus: Array<BakedSku>,
};

export const mockedData: Array<ProductResponse> = [
  {
    id: 1,
    context: {
      name: 'default',
      attributes: {
        language: 'EN',
        modality: 'desktop',
      },
    },
    product: {
      id: 1,
      attributes: {
        title: {
          t: 'string',
          v: 'Flonkey',
        },
        images: {
          t: 'images',
          v: [
            'http://lorempixel.com/75/75/fashion/',
          ],
        },
        description: {
          t: 'string',
          v: 'Best in Class Flonkey',
        },
      },
      variants: {
        default: {},
      },
      skus: {
        default: {
          'SKU-YAX': {},
        },
      },
      activeFrom: '2016-03-14T18:18:47.187Z',
    },
    skus: [
      {
        code: 'SKU-YAX',
        attributes: {
          price: {
            t: 'price',
            v: {
              value: 3300,
              currency: 'USD',
            },
          },
          title: {
            t: 'string',
            v: 'Flonkey',
          },
        },
        activeFrom: '2016-03-14T18:18:47.231Z',
      },
    ],
  },
];
