// modules/product-details.js
/* @flow */

import { createReducer, createAction } from 'redux-act';

import createAsyncActions from './async-utils';

type Attribute = {
  t: string,
  v: any,
};

type Attributes = { [key:string]: Attribute };

type BakedProduct = {
  id: number,
  attributes: Attributes,
  activeFrom?: string,
  activeTo?: string,
};

type BakedSku = {
  code: string,
  attributes: Attributes,
  activeFrom?: string,
  activeTo?: string,
};

type Context = {
  name: string,
  attributes: { [key:string]: string },
};

export type ProductResponse = {
  id: number,
  context: Context,
  product: BakedProduct,
  skus: Array<BakedSku>,
};

function fetchProduct(id: number): global.Promise {
  return this.phoenix.get(`/v1/public/products/${id}`);
}

const {fetch, ...actions} = createAsyncActions('pdp', fetchProduct);

export const resetProduct = createAction('RESET_PRODUCT');

const initialState = {
  product: null,
};

const reducer = createReducer({
  [actions.fetchSucceeded]: (state, payload) => ({
    ...state,
    product: payload,
  }),
  [actions.fetchFailed]: state => ({
    ...state,
    product: null,
  }),
  [resetProduct]: state => ({
    ...state,
    product: null,
  }),
}, initialState);

export {
  fetch,
  reducer as default,
};
