// modules/search.js
/* @flow */

import { get, filter, includes } from 'lodash';
import { createReducer, createAction } from 'redux-act';
import createAsyncActions from './async-utils';
import { mockedData } from './products';
import type { Product } from './products';

export type Search = {
  term: string;
  results: Array<Product>;
}

const INITIAL_STATE:Search = {
  term: '',
  isActive: false,
  results: [],
};

/**
 * Generate search api call actions and reducer
 * Mocked for now
 */
function searchApiCall(searchString: string):Promise {
  let result = mockedData;

  if (searchString != void 0) {
    result = filter(mockedData, (item: Product) => includes(item.name.toLowerCase(), searchString.toLowerCase()));
  }

  return Promise.resolve(result);
}

const { fetch, ...searchActions } = createAsyncActions('search', searchApiCall);

/**
 * External actions
 */
export const setTerm = createAction('SET_TERM');
export const resetTerm = createAction('RESET_TERM');
export const toggleActive = createAction('TOGGLE_ACTIVE');
export { fetch };

const reducer = createReducer({
  [searchActions.fetchSucceeded]: (state, payload) => {
    return {
      ...state,
      results: payload,
    };
  },
  [setTerm]: (state, payload) => {
    return {
      ...state,
      term: payload,
    };
  },
  [resetTerm]: (state) => {
    return {
      ...state,
      term: '',
    };
  },
  [toggleActive]: (state) => {
    return {
      ...state,
      isActive: !get(state, 'isActive', false),
    };
  },
}, INITIAL_STATE);

export default reducer;
