// ui/editable-block/index.jsx

/* @flow */

import React from 'react';
import cssModules from 'react-css-modules';
import styles from './index.css';

import type { HTMLElement } from 'types';

type EditableProps = {
  isEditing: boolean;
  collapsed?: boolean;
  className?: string;
  viewContent: ?HTMLElement|string;
  editContent: ?HTMLElement|string;
  editAction: () => any,
  title: string;
};

const EditableBlock = (props: EditableProps) => {
  const content = props.isEditing ? props.editContent : props.viewContent;

  const editLink = !props.isEditing && !props.collapsed
    ? <div onClick={props.editAction} styleName="edit">EDIT</div>
    : null;

  const contentWrapper = !props.collapsed ? <div styleName="content">{content}</div> : null;

  return (
    <div styleName="editable-block" className={props.className}>
      <div styleName="header">
        <div styleName="title">{props.title}</div>
        {editLink}
      </div>
      {contentWrapper}
    </div>
  );
};

EditableBlock.defaulProps = {
  isEditing: false,
  collapsed: false,
};

export default cssModules(EditableBlock, styles);
