// api-js/lib/endpoints.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var MAX_RESULTS = 1000;

// auth endpoints
var login = exports.login = '/v1/public/login';
var signup = exports.signup = '/v1/public/registrations/new';
var googleSignin = exports.googleSignin = '/v1/public/signin/google/customer';
var logout = exports.logout = '/v1/public/logout';

// product endpoints
var search = exports.search = '/search/products_catalog_view/_search?size=' + MAX_RESULTS;

// cart endpoints
var cart = exports.cart = '/v1/my/cart';
var cartCheckout = exports.cartCheckout = '/v1/my/cart/checkout';
var shippingMethods = exports.shippingMethods = '/v1/my/cart/shipping-methods';
var shippingMethod = exports.shippingMethod = '/v1/my/cart/shipping-method';
var shippingAddress = exports.shippingAddress = '/v1/my/cart/shipping-address';
var shippingAddressId = exports.shippingAddressId = function shippingAddressId(id) {
  return shippingAddress + '/id';
};
var cartLineItems = exports.cartLineItems = '/v1/my/cart/line-items';
var cartPaymentCreditCarts = exports.cartPaymentCreditCarts = '/v1/my/cart/payment-methods/credit-cards';
var cartPaymentGiftCards = exports.cartPaymentGiftCards = '/v1/my/cart/payment-methods/gift-cards';
var cartPaymentStoreCredits = exports.cartPaymentStoreCredits = '/v1/my/cart/payment-methods/store-credits';

var addToCart = exports.addToCart = '/v1/my/cart/add';

var removeFromCart = exports.removeFromCart = '/v1/my/cart/line-items/:id/edit';

// address endpoints
var addresses = exports.addresses = '/v1/my/addresses';
var address = exports.address = function address(addressId) {
  return '/v1/my/addresses/' + addressId;
};
var addressDefault = exports.addressDefault = function addressDefault(addressId) {
  return address(addressId) + '/default';
};
var addressesDefault = exports.addressesDefault = '/v1/my/address/default';

// payment methods, credit cards
var creditCards = exports.creditCards = '/v1/my/payment-methods/credit-cards';
var creditCard = exports.creditCard = function creditCard(creditCardId) {
  return creditCards + '/' + creditCardId;
};
var creditCardDefault = exports.creditCardDefault = function creditCardDefault(creditCardId) {
  return creditCard(creditCardId) + '/default';
};

// payment methods, store credits
var storeCredit = exports.storeCredit = function storeCredit(storeCreditId) {
  return '/v1/my/payment-methods/store-credits/' + storeCreditId;
};
var storeCreditTotals = exports.storeCreditTotals = '/v1/my/payment-methods/store-credits/totals';
var storeCredits = exports.storeCredits = '/search/store_credits_search_view/_search';
