// wings/lib/payment-cards.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.detectCardType = detectCardType;
exports.cardMask = cardMask;
exports.cvvLength = cvvLength;
exports.trimWhiteSpace = trimWhiteSpace;
exports.isCardNumberValid = isCardNumberValid;
exports.isCvvValid = isCvvValid;
function detectCardType(cardNumber) {
  if (cardNumber == void 0) return;

  if (/^3[47]/.test(cardNumber)) {
    return 'amex';
  } else if (/^30[0-5]/.test(cardNumber) || /^3[68]/.test(cardNumber)) {
    return 'dinners-club';
  } else if (/^5[1-5]/.test(cardNumber)) {
    return 'master-card';
  } else if (/^4/.test(cardNumber)) {
    return 'visa';
  }
}

function cardMask(cardType) {
  switch (cardType) {
    case 'amex':
      return '9999 999999 99999';
    case 'dinners-club':
      return '9999 999999 9999';
    default:
      return '9999 9999 9999 9999';
  }
}

function cvvLength(cardType) {
  if (cardType == 'amex') {
    return 4;
  }
  return 3;
}

function trimWhiteSpace(cardNumber) {
  return cardNumber.replace(/\s/g, '');
}

function isCardNumberValid(cardNumber) {
  var mask = cardMask(detectCardType(cardNumber)).replace(/[^\d]/g, '');

  return mask.length === trimWhiteSpace(cardNumber).length;
}

function isCvvValid(cvv, cardType) {
  return cvv.length == cvvLength(cardType);
}
